### ISSUE TITLE
VU Meter: ability to set multiple LEDs per track at once.

### ISSUE THREAD

**Issue by sirwesley** [Aldaviva/BehringerXTouchExtender#8]
Currently, it appears to be possible to only set 1 LED per VU Meter track. 

Maybe it could accept an array of [0...8]?  

Example:  [0, 8] would light Top & Bottom
Example: [0,1,2,3,4] would light Bottom 5.
Etc. 

Note: I'm using the X-Touch Extender as MIDI controller to send to a DMX controller for stage lighting control.  So could be fun to have more control over those LEDs in the VU Meter.

**Comment by Aldaviva** [comment]
Yeah, I really wish we could enable multiple LEDs on the same VU meter at the same time.

Unfortunately, that's not possible on this MIDI device. From the [official documentation](https://mediadl.musictribe.com/media/PLM/data/docs/P0CCR/X-TOUCH%20EXTENDER_QSG_WW.pdf) (PDF page 20, "X-TOUCH EXTENDER MIDI Receive", my emphasis):
> Level Meter Receive value 0...127 (bottom/SIG to top/CLIP, **only 1 LED on at a time**)

The [product shots](https://www.behringer.com/product.html?modelCode=P0CCR) make it look like illuminating multiple VU LEDs possible, but it turns out this only works with non-MIDI protocols like MC (Mackie Control) and HUI (Mackie Human User Interface), which are not how this library communicates with the device and also completely undocumented with no available clients.

There used to be a post on the MusicTribe forums asking this same question, and the answer was basically "it's a protocol-specific capability." Sadly, the forums were deleted a year or two ago and weren't archived.

**Comment by toby11** [comment]
@Aldaviva do you have a copy of the official documentation - that link above seems to be dead now. X-TOUCH%20EXTENDER_QSG_WW.pdf

**Comment by sirwesley** [comment]
https://mediadl.musictribe.com/media/sys_master/hd3/h51/8849820287006.pdf try this?

**Comment by toby11** [comment]
thanks - got it - assume now one has found a way to have all leds lit - I have run the xtouch in reaper in mackie mode and it works that way with displaying more than one led at a time

have you seen this document here where someone seems to have reverse engineered it

http://www.budgetfeatures.com/XctlDOC/Xctl%20Protocol%20for%20X-Touch%20V1.0.pdf

**Comment by toby11** [comment]
![image](https://github.com/user-attachments/assets/db2fba87-369e-44e4-b380-a38c65c7eb8b)
 this looks promising

**Comment by Aldaviva** [comment]
@toby11 thanks for that document, I have not seen that before. It looks like some great reverse engineering work went into that by someone a lot more familiar with MIDI than me.
Also thanks @sirwesley for the working official documentation link. I have checked all the links in the readme and made sure none of them were dead. Also, in case anyone ever wants the originals I linked to above, or if Behringer deletes their website, I saved backup files in [Reference Material](https://github.com/Aldaviva/BehringerXTouchExtender/tree/master/Reference%20Material).

I'll see if I can use that document by FK to get multiple LEDs in one channel's VU meter to illuminate simultaneously by sending MIDI Channel 1 Pressure values. I'm still not sure which of the 4 communication protocols the X-Touch Extender needs to be configured in to receive these inputs; I'm a little worried they're only available in HUI or MC modes, and it seemed like FK was using multiple concurrent connections over different interfaces (USB and MIDI DIN ports simultaneously) for some of this signaling, although that was maybe only required for other features or for tracing the signals. This may work fine in Ctrl and CtrlRel modes (what FK seems to call pure or Xctl) because there is no *1 footnote in the Signal Metering section, but I still don't understand everything in this document and to be honest it's a little confusing.

**Comment by Aldaviva** [comment]
I tried using the MIDI events in the Xctl Protocol document, but none of the VU Meter LEDs lit up. Here is what I attempted, in case anyone sees a mistake or wants to try themselves.

### [`VuMeter.cs`](https://github.com/Aldaviva/BehringerXTouchExtender/blob/master/BehringerXTouchExtender/TrackControls/VuMeter.cs)
```cs
internal override void WriteStateToDevice(object? sender = null, PropertyChangedEventArgs? args = null) {
    ChannelAftertouchEvent pressureEvent = new() {
        Channel         = (FourBitNumber) 1, // I also tried 0
        AftertouchValue = (SevenBitNumber) (TrackId * 16 + LightPosition.Value) // both TrackId and LightPosition.Value are 0-indexed
    };

    _midiClient.AssertOpen();
    _midiClient.ToDevice?.SendEvent(pressureEvent);
}
```
I ran this with the [AudioMeters sample](https://github.com/Aldaviva/BehringerXTouchExtender/blob/master/Sample/AudioMeters.cs). I tested with the X-Touch Extender connected over USB in each of the 4 modes: CtrlRel, Ctrl, MC, and HUI.

**Comment by toby11** [comment]
@Aldaviva thanks from reading that document again I think you may be right that it only works in Xctl mode - I have tested here with your code above and nothing seems to happen. I tested with USB and MIDI din connection.

Going to have a bit more a play with it later though.

**Comment by Aldaviva** [comment]
Thanks, let us know what you find.

**Comment by toby11** [comment]
@Aldaviva I have got multiple leds to show - only in HUI mode though 

see documentation here

https://forum.cockos.com/showthread.php?t=101328

if you look at this

https://stash.reaper.fm/12336/HUIREFTX.txt

sending the raw midi bytes (over USB) works - i.e.  sending these raw bytes resulted in this

`controller.SendMidiMessage(0xa0, 0x00, 0x04);
controller.SendMidiMessage(0xa0, 0x01, 0x05);
                controller.SendMidiMessage(0xa0, 0x02, 0x06);
                controller.SendMidiMessage(0xa0, 0x03, 0x06);
                controller.SendMidiMessage(0xa0, 0x04, 0x07);
                controller.SendMidiMessage(0xa0, 0x05, 0x08);
                controller.SendMidiMessage(0xa0, 0x06, 0x07);
controller.SendMidiMessage(0xa0, 0x07, 0x06);`

![x touch ledsjpg](https://github.com/user-attachments/assets/31895995-22cd-4c0f-a10b-dc6a66526b57)

So there is a way to do it but would involve rewriting your library but to use Mackie HUI mode midi messages as all the other midi messages are different as well. I may fork your project and have a go if thats ok - or if your up for a team effort?

**Comment by Aldaviva** [comment]
Wow impressive! Great work. Feel free to fork this repo and see what you can get working.

On my side, let me start by seeing if I can get anything working with the underlying MIDI library this project uses, and if so I can try to find the time to scaffold out a HUI class and rearrange the class structure to accommodate an additional implementation that integrates with the existing interface, factory, and shares code where possible with the abstract base class.

**Comment by toby11** [comment]
ok great, I am going to try all the other functionality out for the Mackie HUI messages first to confirm the rest of the faders, encoders, scribble strips etc will work as well.

On another note - on my X Touch full (not the extender one) the scribble strips do not seem to work with your code - but not really checked why yet.

**Comment by Aldaviva** [comment]
One potential reason for scribble strips not working on the X-Touch might be a different device ID. The [SysEx message to set the text and color](https://github.com/Aldaviva/BehringerXTouchExtender/blob/ea22ac489cb7f82074563ad93e1f1589fb212543/BehringerXTouchExtender/TrackControls/ScribbleStrip.cs#L42) includes `0x15` at offset 3 (ignore the length comment, that was a coincidence), but the regular X-Touch seems to require `0x14` instead based on [MIDI Mode interface specifications](https://mediadl.musictribe.com/download/software/behringer/X-TOUCH/Document_BE_X-TOUCH-X-TOUCH-EXTENDER-MIDI-Mode-Implementation.pdf). I wonder if there is a way to automatically detect which device is connected, to avoid an extra configuration burden on a user.

**Comment by toby11** [comment]
yes your right - that fixed the scribble strips on mine. thanks

**Comment by toby11** [comment]
from a quick test using the HUI docs above it appears it all pretty much works, I have got everything lit up

![2024-10-10 11 25 07](https://github.com/user-attachments/assets/631aaece-143f-4511-b4b4-4df78e5cf620)

There are three buttons that are not lit but will try them later. 

I think the only limitation will be the scribble strips that you cannot choose the color but that does not bother me really

Events when moving faders and pots all seem to work.

so all looks really promising :)

**Comment by Aldaviva** [comment]
Excellent! That looks really great.

I've been having trouble getting the [DryWetMidi](https://melanchall.github.io/drywetmidi/index.html) library that this project uses to send the HUI VU meter 3-byte events that you showed above, so I will come back and try again later.

**Comment by toby11** [comment]
yeah I normally use NAudio.Midi in my midi projects as I remember having issues wth DryWetMidi

however I am using this project to test it quick

https://github.com/jskeet/DemoCode/tree/main/XTouchMini

the console app lets you send midi bytes from the console to test easily

here is the code to light everything up

```cs
await using (var controller = await XTouchMiniMackieController.ConnectAsync(NullLogger.Instance, "X-Touch"))
    {
    // fader vu meters
    controller.SendMidiMessage(0xa0, 0x00, 0x04);
    controller.SendMidiMessage(0xa0, 0x01, 0x05);
    controller.SendMidiMessage(0xa0, 0x02, 0x06);
    controller.SendMidiMessage(0xa0, 0x03, 0x06);
    controller.SendMidiMessage(0xa0, 0x04, 0x07);
    controller.SendMidiMessage(0xa0, 0x05, 0x08);
    controller.SendMidiMessage(0xa0, 0x06, 0x07);
    controller.SendMidiMessage(0xa0, 0x07, 0x06);                

    // scribble strip "test"
    for (int ifader = 0; ifader < 8; ifader++)
    {
        controller.SendMidiMessage(0xf0, 0x00, 0x00, 0x66, 0x05, 0x00, 0x10, (byte)ifader, 0x74, 0x65, 0x73, 0x74, 0xf7);
    }
    
    // time code
    controller.SendMidiMessage(0xf0, 0x00, 0x00, 0x66, 0x05, 0x00, 0x11, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0xf7);

    // v pots
    int vpot = 20;
    for (int ifader = 0x10; ifader < 0x19; ifader++)
    {
        controller.SendMidiMessage(0xb0, (byte)ifader, (byte)vpot++);
    }

     // all led zones
    for(int iZone =0; iZone < 0x1e; iZone++)
    {
        // set zone first
        controller.SendMidiMessage(0xb0, 0x0c, (byte)iZone);

        // then send bytes
        controller.SendMidiMessage(0xb0, 0x2c, 0x40);
        controller.SendMidiMessage(0xb0, 0x2c, 0x41); 
        controller.SendMidiMessage(0xb0, 0x2c, 0x42);
        controller.SendMidiMessage(0xb0, 0x2c, 0x43);
        controller.SendMidiMessage(0xb0, 0x2c, 0x44);
        controller.SendMidiMessage(0xb0, 0x2c, 0x45);
        controller.SendMidiMessage(0xb0, 0x2c, 0x46);
        controller.SendMidiMessage(0xb0, 0x2c, 0x47);                                       
    }
}
```

**Comment by Aldaviva** [comment]
Nice, I'm a little bit familiar with NAudio from using their WASAPI package for audio file playback once.

So far it seems like DryWetMidi doesn't make it possible to send custom event bytes, all the classes have internal methods I can't override, and if my class doesn't inherit from specific well-known superclasses, my event is silently ignored. So it seems to not be very extensible, unfortunately.

I'll be signing off for now and will try to come back to this some time later. Thanks for all your effort on this.

**Comment by Aldaviva** [comment]
I figured out how to set the VU meters in HUI mode using DryWetMidi, and I learned three important facts.

1. The Xctl protocol document by FK is wrong, at least for my X-Touch Extender in HUI mode: you have to send **_Note_** Aftertouch/Polyphonic Key Pressure events, not **_Channel_** Aftertouch/Pressure events, which are different. I thought I had tried both earlier but I must have gotten the parameters wrong on the correct event.
    - I missed this initially because the NoteAftertouch was incorrectly shown to have a status byte of `0x0a`, so I thought AgeMan's byte array that didn't start with `0x0a` wasn't a Note Aftertouch event.
    - What I now know is that the idea of a MIDI status byte is a lie, it's actually 4 bits of status, not 8, and it's bit-shifted and bit-packed with the channel number to form the real leading byte.
    - A correct way to have represented this would have been as either 4-bit numbers, or already shifted (like `0xa0`) to be ANDed with the channel number.
1. The aftertouch event note # is the 0-indexed track ID [0, 7], and the aftertouch value is a non-linear conversion of the number of LEDs you want. See **Code sample** below for one implementation of a switch statement that simplifies this logic. Alternatively, a lookup table would also work.
    |# of LEDs to illuminate|Aftertouch value|
    |-:|-:|
    |0|0|
    |1|1 or 2|
    |2|3|
    |3|4|
    |4|5|
    |5|6 or 7|
    |6|8 or 9|
    |7|10 or 11|
    |8|12 or more|
1. If you know the bytes you want to send but aren't sure what MIDI event they correspond to, you can use `Melanchall.DryWetMidi.Core.BytesToMidiEventConverter.Convert(byte[])` to parse them into a `MidiEvent`, which is how I figured out the `NoteAftertouchEvent` fact above.
    - I am not sure if this can also send arbitrary byte array payloads that don't conform to any known MIDI events, but it's at least supposed to be able to, which would be useful.
        ```cs
        outputDevice.SendEvent(bytesToMidiEventConverter.Convert([0xa0, 0x00, 0x0c]));
        ```
    - This behavior is mostly undocumented, but does show up in [one issue](https://github.com/melanchall/drywetmidi/issues/55#issuecomment-567432218) that I eventually found 6 pages deep in the issue tracker.

### Code sample
```cs
public void SetVuMeter(int trackIndex, int level) {
    SevenBitNumber aftertouchValue = (SevenBitNumber) (level switch {
        0    => 0,
        <= 5 => level + 1,
        _    => 2 * (level - 6) + 8
    });
    MidiEvent midiEvent = new NoteAftertouchEvent((SevenBitNumber) trackIndex, aftertouchValue);
    _midiClient.ToDevice?.SendEvent(midiEvent);
}
```

### Photo
![PXL_20241011_062820851 (2)](https://github.com/user-attachments/assets/ec706f9e-36ec-4462-8527-c9562265a16c)

**Comment by toby11** [comment]
Nice - I have not really used the xctl document - I have just been going on what it says in the HUI documentation I found - also from testing yesterday the only other issue I found on mine (which won't affect the extender) is that you cannot use the main fader to control anything as this is not supported in HUI mode

found some detail about it here

https://community.musictribe.com/discussion/solutions/10/0/251752/x-touch-and-pro-tools

mainly this comment here

_I'm told this is because HUI only implements 8 faders in it's control protocol and is not a problem with your X-TOUCH controller. Because the Main fader is the 9th one this will not be recognized sing HUI._

**Comment by Aldaviva** [comment]
I managed to organize the codebase in a way that allows for reuse of shared logic, extensibility for HUI implementations of different controls (currently mostly unimplemented), and API compatibility and interchangeability for library consumers.

I'll check this into the [`hui`](https://github.com/Aldaviva/BehringerXTouchExtender/tree/hui) branch. Here are the [VU meters](https://github.com/Aldaviva/BehringerXTouchExtender/blob/ddc16915d9287414957245ad9e9c50784008389d/BehringerXTouchExtender/TrackControls/Hui/HuiVuMeter.cs) working in that branch.

https://github.com/user-attachments/assets/e035d5b2-64dc-4a43-b5d9-7c845c93bce4

**Comment by sirwesley** [comment]
Been distracted with other projects, but awesome work getting the VU meters lighting up.   Should we update the ReadMe to reflect this advancement?   Give it the ol' AI once over to maybe get another release published?  I've got some projects coming up that will want to dig into this again soon, so just curious where we've left things.

**Comment by Aldaviva** [comment]
Yeah I can work on getting this released. I believe the remaining effort was mostly documentation and testing, but I think the code was all there and functional when I last left it.

**Comment by Aldaviva** [comment]
Deployed [2.0.0-beta.0](https://www.nuget.org/packages/BehringerXTouchExtender/2.0.0-beta.0) to NuGet Gallery. May still need more manual testing, but all automated tests are now passing.

**Comment by Aldaviva** [comment]
Deployed [2.0.0](https://www.nuget.org/packages/BehringerXTouchExtender/2.0.0) release to NuGet Gallery.
